source/gui/settingsDialogs.py

```diff
@@ -2879,6 +3074,392 @@ def onBlinkCursorChange(self, evt):
+			self,
+			parent: wx.Window,
+			*,  # Make next argument keyword only
+			settingsCallable: Callable[[], vision.providerBase.VisionEnhancementProviderSettings]
+	):
+		"""
+		@param settingsCallable: A callable that returns an instance to a VisionEnhancementProviderSettings.
+			This will usually be a weakref, but could be any callable taking no arguments.
+		"""
+		self._settingsCallable = settingsCallable
+		super().__init__(parent=parent)
+
+	def getSettings(self) -> AutoSettings:
+		settings = self._settingsCallable()
+		return settings
+
+	def makeSettings(self, settingsSizer):
+		# Construct vision enhancement provider settings
+		self.updateDriverSettings()
+
+
+class VisionProviderSubPanel_Wrapper(
+		SettingsPanel
+):
+
+	_checkBox: wx.CheckBox
+
+	def __init__(
+			self,
+			parent: wx.Window,
+			providerControl: VisionProviderStateControl
+	):
+		self._providerControl = providerControl
+		self._providerSettings: Optional[VisionProviderSubPanel_Settings] = None
+		self._providerSettingsSizer = wx.BoxSizer(orient=wx.VERTICAL)
+		super().__init__(parent=parent)
+
+	def makeSettings(self, settingsSizer):
+		# Translators: Enable checkbox on a vision enhancement provider on the vision settings category panel
+		checkBox = wx.CheckBox(self, label=_("Enable"))
+		settingsSizer.Add(checkBox)
+		settingsSizer.AddSpacer(size=self.scaleSize(10))
+		# Translators: Options label on a vision enhancement provider on the vision settings category panel
+		settingsSizer.Add(wx.StaticText(self, label=_("Options:")))
+		settingsSizer.Add(
+			self._providerSettingsSizer,
+			border=self.scaleSize(15),
+			flag=wx.LEFT | wx.EXPAND,
+			proportion=1.0
+		)
+		self._checkBox: wx.CheckBox = checkBox
+		if self._providerControl.getProviderInstance():
+			checkBox.SetValue(True)
+		if self._createProviderSettings():
+			checkBox.Bind(wx.EVT_CHECKBOX, self._enableToggle)
+		else:
+			checkBox.Bind(wx.EVT_CHECKBOX, self._nonEnableableGUI)
+
+	def _createProviderSettings(self):
+		try:
+			getSettingsCallable = self._providerControl.getProviderInfo().providerClass.getSettings
+			self._providerSettings = VisionProviderSubPanel_Settings(
+				self,
+				settingsCallable=getSettingsCallable
+			)
+			self._providerSettingsSizer.Add(self._providerSettings, flag=wx.EXPAND, proportion=1.0)
+		# Broad except used since we can not know what exceptions a provider might throw.
+		# We should be able to continue despite a buggy provider.
+		except Exception:
+			log.error("unable to create provider settings", exc_info=True)
+			return False
+		return True
+
+	def _nonEnableableGUI(self, evt):
+		wx.MessageBox(
+			# Translators: Shown when there is an error showing the GUI for a vision enhancement provider
+			_("Unable to configure user interface for Vision Enhancement Provider, it can not be enabled."),
+			parent=self,
+		)
+		self._checkBox.SetValue(False)
+
+	def _enableToggle(self, evt):
+		if evt.IsChecked():
+			self._providerControl.startProvider()

```

Comment:
If this fails, the check box still remains checked. Therefore, if enabling the check box results in an error, it should be disabled again and we should refrain from updating the gui.

Reply:
Good catch